### PR TITLE
Zoom to features with a max scale constraint

### PR DESCRIPTION
**Pull request by nboisteault**

### CODE REVIEW

File: assets/src/modules/LocateByLayer.js [hunk 1]
```diff
@@ -0,0 +1,472 @@
...
+
+/**
+ * @class
+ * @name LocateByLayer
+ */
+export default class LocateByLayer {
+    constructor() {
+        this._locateByLayerConfig = lizMap.config.locateByLayer;
+        const locateBtn = document.getElementById('button-locate');
+        if (mainLizmap.initialConfig.locateByLayer) {
```
Comment (rldhont) [inline assets/src/modules/LocateByLayer.js:22]:
```suggestion
    /**
     * Build the lizmap LocateByLayer instance
     * @param {LocateByLayerConfig} locateByLayer - The lizmap locateByLayer config
     * @param {WfsFeatureType[]} vectorLayerFeatureTypeList - The list of WFS feature type
     * @param {Map}           map           - OpenLayers map
     * @param {object}        lizmap3       - The old lizmap object
     */
    constructor(locateByLayer, vectorLayerFeatureTypeList, map, lizmap3) {
        this._map = map;
        this._vectorLayerFeatureTypeList = vectorLayerFeatureTypeList;
        
        this._lizmap3 = lizmap3;
        this._lizmap3LocateByLayerConfig = lizmap3.config.locateByLayer;
        
        const locateBtn = document.getElementById('button-locate');
        if (locateByLayer) {
```

File: assets/src/modules/LocateByLayer.js [hunk 2]
```diff
@@ -0,0 +1,472 @@
...
+        } else {
+            locateBtn?.parentNode.classList.add('hide');
+        }
+    }
+
+    addLocateByLayer() {
+        var locateByLayerList = [];
+        for (var lname in this._locateByLayerConfig) {
+            if ('order' in this._locateByLayerConfig[lname])
+                locateByLayerList[this._locateByLayerConfig[lname].order] = lname;
```
Comment (rldhont) [inline assets/src/modules/LocateByLayer.js:35]:
```suggestion
        for (var lname in this._lizmap3LocateByLayerConfig) {
            if ('order' in this._lizmap3ocateByLayerConfig[lname])
                locateByLayerList[this._lizmap3ocateByLayerConfig[lname].order] = lname;
```

File: assets/src/modules/LocateByLayer.js [hunk 3]
```diff
@@ -0,0 +1,472 @@
...
+            else
+                locateByLayerList.push(lname);
+        }
+        var locateContent = [];
+        for (var l in locateByLayerList) {
+            var lname = locateByLayerList[l];
+            var lConfig = lizMap.config.layers[lname];
```
Comment (rldhont) [inline assets/src/modules/LocateByLayer.js:42]:
```suggestion
            var lConfig = this._lizmap3.config.layers[lname];
```

File: assets/src/modules/LocateByLayer.js [hunk 4]
```diff
@@ -0,0 +1,472 @@
...
+            var html = '<div class="locate-layer">';
+            html += '<select id="locate-layer-' + lizMap.cleanName(lname) + '" class="label">';
```
Comment (rldhont) [inline assets/src/modules/LocateByLayer.js:44]:
```suggestion
            html += '<select id="locate-layer-' + this._lizmap3.cleanName(lname) + '" class="label">';
```

File: assets/src/modules/LocateByLayer.js [hunk 5]
```diff
@@ -0,0 +1,472 @@
...
+            html += '</select>';
+            html += '</div>';
+            //constructing the select
+            locateContent.push(html);
+        }
+        $('#locate .menu-content').html(locateContent.join('<hr/>'));
+
+        var featureTypes = lizMap.mainLizmap.initialConfig.vectorLayerFeatureTypeList;
+        if (featureTypes.length == 0) {
+            this._locateByLayerConfig = {};
```
Comment (rldhont) [inline assets/src/modules/LocateByLayer.js:55]:
```suggestion
            this._lizmap3LocateByLayerConfig = {};
```

File: assets/src/modules/LocateByLayer.js [hunk 6]
```diff
@@ -0,0 +1,472 @@
...
+            for (const featureType of featureTypes) {
+                var typeName = featureType.Name;
+                var lname = lizMap.getNameByTypeName(typeName);
+                if (!lname) {
+                    if (typeName in this._locateByLayerConfig)
+                        lname = typeName
+                    else if ((typeName in shortNameMap) && (shortNameMap[typeName] in this._locateByLayerConfig))
+                        lname = shortNameMap[typeName];
+                    else {
+                        for (var lbl in this._locateByLayerConfig) {
```
Comment (rldhont) [inline assets/src/modules/LocateByLayer.js:68]:
```suggestion
                    if (typeName in this._lizmap3LocateByLayerConfig)
                        lname = typeName
                    else if ((typeName in shortNameMap) && (shortNameMap[typeName] in this._lizmap3LocateByLayerConfig))
                        lname = shortNameMap[typeName];
                    else {
                        for (var lbl in this._lizmap3LocateByLayerConfig) {
```

File: assets/src/modules/LocateByLayer.js [hunk 7]
```diff
@@ -0,0 +1,472 @@
...
+                                break;
+                            }
+                        }
+                    }
+                }
+
+                if (!(lname in this._locateByLayerConfig))
+                    continue;
+
+                var locate = this._locateByLayerConfig[lname];
```
Comment (rldhont) [inline assets/src/modules/LocateByLayer.js:80]:
```suggestion
                if (!(lname in this._lizmap3LocateByLayerConfig))
                    continue;

                var locate = this._lizmap3LocateByLayerConfig[lname];
```

File: assets/src/modules/LocateByLayer.js [hunk 8]
```diff
@@ -0,0 +1,472 @@
...
+            }
+
+            // get joins
+            for (var lName in this._locateByLayerConfig) {
+                var locate = this._locateByLayerConfig[lName];
+                if ('vectorjoins' in locate && locate['vectorjoins'].length != 0) {
+                    var vectorjoin = locate['vectorjoins'][0];
+                    locate['joinFieldName'] = vectorjoin['targetFieldName'];
+                    for (var jName in this._locateByLayerConfig) {
+                        var jLocate = this._locateByLayerConfig[jName];
```
Comment (rldhont) [inline assets/src/modules/LocateByLayer.js:95]:
```suggestion
            for (var lName in this._lizmap3LocateByLayerConfig) {
                var locate = this._lizmap3LocateByLayerConfig[lName];
                if ('vectorjoins' in locate && locate['vectorjoins'].length != 0) {
                    var vectorjoin = locate['vectorjoins'][0];
                    locate['joinFieldName'] = vectorjoin['targetFieldName'];
                    for (var jName in this._lizmap3LocateByLayerConfig) {
                        var jLocate = this._lizmap3LocateByLayerConfig[jName];
```

File: assets/src/modules/LocateByLayer.js [hunk 9]
```diff
@@ -0,0 +1,472 @@
...
+                                'joinLayerId': locate.layerId,
+                                'joinLayer': lName
+                            }];
+                        }
+                    }
+                }
+            }
+
+            // get locate by layers features
+            for (var lname in this._locateByLayerConfig) {
```
Comment (rldhont) [inline assets/src/modules/LocateByLayer.js:113]:
```suggestion
            for (var lname in this._lizmap3LocateByLayerConfig) {
```

File: assets/src/modules/LocateByLayer.js [hunk 10]
```diff
@@ -0,0 +1,472 @@
...
+                this.getLocateFeature(lname);
+            }
+            $('.btn-locate-clear').click(function () {
+                lizMap.mainLizmap.map.clearHighlightFeatures();
+                $('#locate select').val('-1');
+                $('div.locate-layer span > input').val('');
+
+                if (lizMap.lizmapLayerFilterActive) {
+                    lizMap.events.triggerEvent('lizmaplocatefeaturecanceled',
```
Comment (rldhont) [inline assets/src/modules/LocateByLayer.js:122]:
```suggestion
                if (this._lizmap3.lizmapLayerFilterActive) {
                    this._lizmap3.events.triggerEvent('lizmaplocatefeaturecanceled',
```

File: assets/src/modules/LocateByLayer.js [hunk 11]
```diff
@@ -0,0 +1,472 @@
...
+            });
+        }
+    }
+
+     /**
+     * Get features for locate by layer tool
+     * @param aName
+     */
+     getLocateFeature(aName) {
+        var locate = this._locateByLayerConfig[aName];
```
Comment (rldhont) [inline assets/src/modules/LocateByLayer.js:142]:
```suggestion
        var locate = this._lizmap3LocateByLayerConfig[aName];
```

File: assets/src/modules/LocateByLayer.js [hunk 12]
```diff
@@ -0,0 +1,472 @@
...
+        if ( 'vectorjoins' in locate ) {
+            var vectorjoins = locate.vectorjoins;
+            for ( var i=0, len=vectorjoins.length; i<len; i++) {
+                var vectorjoin = vectorjoins[i];
+                fields.push( vectorjoin.targetFieldName );
+            }
+        }
+
+        // Get WFS url and options
+        var getFeatureUrlData = lizMap.getVectorLayerWfsUrl( aName, null, null, 'extent' );
```
Comment (rldhont) [inline assets/src/modules/LocateByLayer.js:166]:
```suggestion
        var getFeatureUrlData = this._lizmap3.getVectorLayerWfsUrl( aName, null, null, 'extent' );
```

File: assets/src/modules/LocateByLayer.js [hunk 13]
```diff
@@ -0,0 +1,472 @@
...
+        getFeatureUrlData['options']['PROPERTYNAME'] = fields.join(',');
+
+        var layerName = lizMap.cleanName(aName);
```
Comment (rldhont) [inline assets/src/modules/LocateByLayer.js:169]:
```suggestion
        var layerName = this._lizmap3.cleanName(aName);
```

File: assets/src/modules/LocateByLayer.js [hunk 14]
```diff
@@ -0,0 +1,472 @@
...
+
+        // Get data
+        $.post( getFeatureUrlData['url'], getFeatureUrlData['options'], (data) => {
+            var lConfig = lizMap.config.layers[aName];
```
Comment (rldhont) [inline assets/src/modules/LocateByLayer.js:173]:
```suggestion
            var lConfig = this._lizmap3.config.layers[aName];
```

File: assets/src/modules/LocateByLayer.js [hunk 15]
```diff
@@ -0,0 +1,472 @@
...
+                var val = event.target.value;
+                if (val == '-1') {
+                    $('#locate-layer-'+layerName+' ~ span > input').val('');
+                    // update to join layer
+                    if ( 'filterjoins' in locate && locate.filterjoins.length != 0 ) {
+                        var filterjoins = locate.filterjoins;
+                        for (var i=0, len=filterjoins.length; i<len; i++) {
+                            var filterjoin = filterjoins[i];
+                            var jName = filterjoin.joinLayer;
+                            if ( jName in this._locateByLayerConfig ) {
```
Comment (rldhont) [inline assets/src/modules/LocateByLayer.js:299]:
```suggestion
                            if ( jName in this._lizmap3LocateByLayerConfig ) {
```

File: assets/src/modules/LocateByLayer.js [hunk 16]
```diff
@@ -0,0 +1,472 @@
...
+                                this.updateLocateFeatureList( jName );
+                                $('#locate-layer-'+cleanName(jName)).val( oldVal );
+                                return;
+                            }
+                        }
+                    }
+                    // zoom to parent selection
+                    if ( 'vectorjoins' in locate && locate.vectorjoins.length == 1 ) {
+                        var jName = locate.vectorjoins[0].joinLayer;
+                        if ( jName in this._locateByLayerConfig ) {
```
Comment (rldhont) [inline assets/src/modules/LocateByLayer.js:311]:
```suggestion
                        if ( jName in this._lizmap3LocateByLayerConfig ) {
```

File: assets/src/modules/LocateByLayer.js [hunk 17]
```diff
@@ -0,0 +1,472 @@
...
+                } else {
+                    // zoom to val
+                    this.zoomToLocateFeature( aName );
+                    // update joined layer
+                    if ( 'filterjoins' in locate && locate.filterjoins.length != 0 ) {
+                        var filterjoins = locate.filterjoins;
+                        for (var i=0, len=filterjoins.length; i<len; i++) {
+                            var filterjoin = filterjoins[i];
+                            var jName = filterjoin.joinLayer;
+                            if ( jName in this._locateByLayerConfig ) {
```
Comment (rldhont) [inline assets/src/modules/LocateByLayer.js:327]:
```suggestion
                            if ( jName in this._lizmap3LocateByLayerConfig ) {
```

File: assets/src/modules/LocateByLayer.js [hunk 18]
```diff
@@ -0,0 +1,472 @@
...
+        // clear highlight layer
+        lizMap.mainLizmap.map.clearHighlightFeatures();
+
+        // get locate by layer val
+        var locate = this._locateByLayerConfig[aName];
+        var layerName = lizMap.cleanName(aName);
+        var val = $('#locate-layer-'+layerName).val();
+        if (val == '-1') {
+            // Trigger event
+            lizMap.events.triggerEvent('lizmaplocatefeaturecanceled', {'featureType': aName });
```
Comment (rldhont) [inline assets/src/modules/LocateByLayer.js:380]:
```suggestion
        var locate = this._lizmap3LocateByLayerConfig[aName];
        var layerName = this._lizmap3.cleanName(aName);
        var val = $('#locate-layer-'+layerName).val();
        if (val == '-1') {
            // Trigger event
            this._lizmap3.events.triggerEvent('lizmaplocatefeaturecanceled', {'featureType': aName });
```

File: assets/src/modules/LocateByLayer.js [hunk 19]
```diff
@@ -0,0 +1,472 @@
...
+                    });
+                }
+                // zoom to extent
+                lizMap.mainLizmap.map.zoomToGeometryOrExtent(geom);
+            }
+
+            var fid = val.split('.')[1];
+
+            // Trigger event
+            lizMap.events.triggerEvent('lizmaplocatefeaturechanged',
```
Comment (rldhont) [inline assets/src/modules/LocateByLayer.js:411]:
```suggestion
            this._lizmap3.events.triggerEvent('lizmaplocatefeaturechanged',
```

File: assets/src/modules/LocateByLayer.js [hunk 20]
```diff
@@ -0,0 +1,472 @@
...
+            );
+        }
+    }
+
+    /**
+     * Get features for locate by layer tool
+     * @param aName
+     */
+    updateLocateFeatureList(aName) {
+        var locate = this._locateByLayerConfig[aName];
```
Comment (rldhont) [inline assets/src/modules/LocateByLayer.js:425]:
```suggestion
        var locate = this._lizmap3LocateByLayerConfig[aName];
```

File: assets/src/modules/LocateByLayer.js [hunk 21]
```diff
@@ -0,0 +1,472 @@
...
+        // clone features reference
+        var features = {};
+        for ( var fid in locate.features ) {
+            features[fid] = locate.features[fid];
+        }
+        // filter by filter field name
+        if ('filterFieldName' in locate) {
+            var filterValue = $('#locate-layer-' + lizMap.cleanName(aName) + '-'+locate.filterFieldName).val();
```
Comment (rldhont) [inline assets/src/modules/LocateByLayer.js:433]:
```suggestion
            var filterValue = $('#locate-layer-' + this._lizmap3.cleanName(aName) + '-'+locate.filterFieldName).val();
```

File: assets/src/modules/LocateByLayer.js [hunk 22]
```diff
@@ -0,0 +1,472 @@
...
+        }
+        // filter by vector joins
+        if ( 'vectorjoins' in locate && locate.vectorjoins.length != 0 ) {
+            var vectorjoins = locate.vectorjoins;
+            for ( var i=0, len =vectorjoins.length; i< len; i++) {
+                var vectorjoin = vectorjoins[i];
+                var jName = vectorjoin.joinLayer;
+                if ( jName in this._locateByLayerConfig ) {
+                    var jLocate = this._locateByLayerConfig[jName];
+                    var jVal = $('#locate-layer-' + lizMap.cleanName(jName)).val();
```
Comment (rldhont) [inline assets/src/modules/LocateByLayer.js:451]:
```suggestion
                    var jLocate = this._lizmap3LocateByLayerConfig[jName];
                    var jVal = $('#locate-layer-' + this._lizmap3.cleanName(jName)).val();
```

File: assets/src/modules/LocateByLayer.js [hunk 23]
```diff
@@ -0,0 +1,472 @@
...
+        }
+        // create the option list
+        const placeHolder = lizMap.config.layers[aName].title;
+        var options = '<option value="-1" label="'+placeHolder+'"></option>';
+        for (var fid in features) {
+            var feat = features[fid];
+            options += '<option value="' + feat.id + '">' + DOMPurify.sanitize(feat.properties[locate.fieldName]) + '</option>';
+        }
+        // add option list
+        $('#locate-layer-'+ lizMap.cleanName(aName)).html(options);
```
Comment (rldhont) [inline assets/src/modules/LocateByLayer.js:470]:
```suggestion
        $('#locate-layer-'+ this._lizmap3.cleanName(aName)).html(options);
```

File: assets/src/modules/LocateByLayer.js [hunk 24]
```diff
@@ -0,0 +1,472 @@
...
+                    for (var fid in features) {
+                        var feat = features[fid];
+                        if ( feat.properties[vectorjoin.targetFieldName] != jFeat.properties[vectorjoin.joinFieldName] )
+                            delete features[fid];
+                    }
+                }
+            }
+        }
+        // create the option list
+        const placeHolder = lizMap.config.layers[aName].title;
```
Comment (rldhont) [inline assets/src/modules/LocateByLayer.js:463]:
```suggestion
        const placeHolder = this._lizmap3.config.layers[aName].title;
```

File: assets/src/modules/LocateByLayer.js [hunk 25]
```diff
@@ -0,0 +1,472 @@
...
+                    }
+                }
+            }
+
+            // get locate by layers features
+            for (var lname in this._locateByLayerConfig) {
+                this.getLocateFeature(lname);
+            }
+            $('.btn-locate-clear').click(function () {
+                lizMap.mainLizmap.map.clearHighlightFeatures();
```
Comment (rldhont) [inline assets/src/modules/LocateByLayer.js:117]:
```suggestion
                this._map.clearHighlightFeatures();
```

File: assets/src/modules/LocateByLayer.js [hunk 26]
```diff
@@ -0,0 +1,472 @@
...
+        },'json');
+    }
+
+    /**
+     * Zoom to locate feature
+     * @param aName
+     */
+    zoomToLocateFeature(aName) {
+        // clear highlight layer
+        lizMap.mainLizmap.map.clearHighlightFeatures();
```
Comment (rldhont) [inline assets/src/modules/LocateByLayer.js:372]:
```suggestion
        this._map.clearHighlightFeatures();
```

File: assets/src/modules/LocateByLayer.js [hunk 27]
```diff
@@ -0,0 +1,472 @@
...
+                        if ( !data.features ){
+                            data = JSON.parse(data);
+                        }
+                        lizMap.mainLizmap.map.setHighlightFeatures(data.features[0], "geojson");
+                    }).fail(function(){
+                        lizMap.mainLizmap.map.setHighlightFeatures(feat, "geojson");
+                    });
+                }
+                // zoom to extent
+                lizMap.mainLizmap.map.zoomToGeometryOrExtent(geom);
```
Comment (rldhont) [inline assets/src/modules/LocateByLayer.js:405]:
```suggestion
                        this._map.setHighlightFeatures(data.features[0], "geojson");
                    }).fail(function(){
                        this._map.setHighlightFeatures(feat, "geojson");
                    });
                }
                // zoom to extent
                this._map.zoomToGeometryOrExtent(geom);
```

File: assets/src/modules/LocateByLayer.js [hunk 28]
```diff
@@ -0,0 +1,472 @@
...
+            html += '<select id="locate-layer-' + lizMap.cleanName(lname) + '" class="label">';
+            html += '<option>' + lConfig.title + '...</option>';
+            html += '</select>';
+            html += '</div>';
+            //constructing the select
+            locateContent.push(html);
+        }
+        $('#locate .menu-content').html(locateContent.join('<hr/>'));
+
+        var featureTypes = lizMap.mainLizmap.initialConfig.vectorLayerFeatureTypeList;
```
Comment (rldhont) [inline assets/src/modules/LocateByLayer.js:53]:
```suggestion
        var featureTypes = this._vectorLayerFeatureTypeList;
```

File: assets/src/modules/Lizmap.js [hunk 29]
```diff
@@ -167,6 +168,7 @@ export default class Lizmap {
                 this.legend = new Legend();
                 this.search = new Search();
                 this.tooltip = new Tooltip();
+                this.locateByLayer = new LocateByLayer();
```
Comment (rldhont) [inline assets/src/modules/Lizmap.js:171]:
```suggestion
                this.locateByLayer = new LocateByLayer(
                    this.initialConfig.locateByLayer,
                    this.initialConfig.vectorLayerFeatureTypeList,
                    this.map,
                    this._lizmap3
                );
```

**Review by rldhont — CHANGES_REQUESTED** [review]
remove `lizMap` from `LocateByLayer` module.

**Comment by nboisteault** [comment]
ok to merge?

**Review by rldhont — APPROVED** [review]

**Comment by rldhont** [comment]
Thanks @nboisteault

**Comment by Gustry** [comment]
Please provide a description in the PR ;-)

**Pull request merged**